execution-engine/src/fs.rs

```diff
@@ -998,18 +993,27 @@ impl FileSystem {
     /// how a particular execution engine handles the memory.
     /// That is, different engines provide different API to interact the linear memory
     /// space of WASM. Hence, the method here return the read bytes as `Vec<u8>`.
-    pub(crate) fn fd_pread(
+    pub(crate) fn fd_pread<B: AsMut<[u8]>>(
         &self,
         fd: Fd,
-        buffer_len: usize,
+        bufs: &mut [B],
         offset: FileSize,
-    ) -> FileSystemResult<Vec<u8>> {
+    ) -> FileSystemResult<usize> {
         self.check_right(&fd, Rights::FD_READ)?;
         let inode = self.fd_table.get(&fd).ok_or(ErrNo::BadF)?.inode;
 
-        self.lock_inode_table()?
-            .get(&inode)?
-            .read_file(buffer_len, offset)
+        let f = self.lock_inode_table()?;
+        let f = f.get(&inode)?;
+
+        let mut offset = offset;
```

Comment:
You can mark the parameter `offset` as `mut` here, instead.

Reply:
You can but that mixes implementation details with the function signature which IMO isn't great.

Here's a really terrible/ridiculous example of a function that does not take a mutable reference:
``` rust
fn do_the_thing(mut x: &[u8]) {
    ...
}
```

Happy to move the mut to the arguments if you feel strongly but I tend to re-declare locals as mut in the function for that reason.

Reply:
No, I think that makes sense.  I'll mark this and the other comment as resolved.